###Mouse Development and Aging 
###File 2
###Gene Expression Analysis
###Last updated Oct 7 2024


```{r}
library(Seurat)
library(SeuratObject) 
library(tidyverse)
library(patchwork)
library(dplyr)
library(tidyr)
library(ggplot2)

#The following is for Volcano Plots. Run install once.
#BiocManager::install("ComplexHeatmap") #required for scRNAtoolVis
#install.packages('devtools')
#devtools::install_github('junjunlab/scRNAtoolVis')

library(scRNAtoolVis) #Volcano Plots; requires tidyverse

#The following is for Gene Ontology; install once
#if (!require("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("clusterProfiler")
library(clusterProfiler) #enrichGO
#BiocManager::install("org.Mm.eg.db")
library(org.Mm.eg.db) #gene ontology
library(enrichplot) #gene ontology
#BiocManager::install("EnrichedHeatmap")
library(EnrichedHeatmap)
#BiocManager::install("rrvgo")
library(rrvgo) #gene ontology

library(viridis) #visibly accessible color profiles
#https://www.data-to-viz.com/graph/heatmap.html
library(heatmaply)

library(ggupset) #upset plot
library(pheatmap)

load("MDA_merge.obj.Rdata")
#load merge.obj from File 1: MDA_1_load_thru_cluster.Rmd

```

### Step 1 - Marker genes by sample
Certain marker genes are expected to be distributed similarly in any analyzed sample. We will check this as a form of quality control.

 
```{r}
color_palette <- c("mf_p21_rep1" = "red", "mf_p21_rep2" = "yellow", "adult_1" = "blue", "adult_2" = "green")
#This ensures the samples are coded with the same colors consistently across our visualizations.

sample_order <- c("mf_p21_rep1", "mf_p21_rep2", "adult_1", "adult_2")
merge.obj$orig.ident <- factor(merge.obj$orig.ident, levels = sample_order)
#This is the order the samples should display from left to right

VlnPlot(merge.obj, features = c("Actb"), group.by = "orig.ident", pt.size = 0) + scale_fill_manual(values = color_palette) + guides(fill = FALSE) 
#If we don't manually define the order, it orders from left to right in ascending order
ggsave("./Figures/Genes_Violin/actb_4_samples.tiff", device='tiff', dpi=500, height = 5, width = 5, unit = 'in')
#automatically saves the image

VlnPlot(merge.obj, features = c("Top1"), group.by = "orig.ident", pt.size = 0) + scale_fill_manual(values = color_palette) + guides(fill = FALSE) 
ggsave("./Figures/Genes_Violin/top1_4_samples.tiff", device='tiff', dpi=500, height = 5, width = 5, unit = 'in')

VlnPlot(merge.obj, features = c("B2m"), group.by = "orig.ident", pt.size = 0) + scale_fill_manual(values = color_palette) + guides(fill = FALSE) 
ggsave("./Figures/Genes_Violin/B2m_4_samples.tiff", device='tiff', dpi=500, height = 5, width = 5, unit = 'in')

```
### Step 2a - Marker Genes by Cluster - avg log2FC
Our goal is to identify genes that are highly expressed in one cluster but poorly expressed in other clusters to serve as marker genes per cluster. This reinforces that each cluster represents distinct anatomical regions that in turn express distinct genes. 
In our lab's FA Project, this was done by considering avg_log2FC. First, we generate a heat map; then we generate violin plots for each cluster using the gene with the highest avg_log2FC. However, that did not produce good results for this dataset using Seurat clusters. The top genes by avg_log2FC in Cluster 0 were also highly expressed in the other clusters. Thus, in Steps 2b & 2c, we consider alternative methods for selecting marker genes by cluster.

```{r}
#Top 10 genes per cluster by average log2FC
merge.obj.marker <- FindAllMarkers(merge.obj, only.pos = TRUE, min.pct = 0.5, logfc.threshold = 0.25)

merge.obj.marker %>%
    group_by(cluster) %>%
    top_n(n = 10, wt = avg_log2FC) -> top10

top10 <- top10[c(1:30,51:60,31:50),]

DoHeatmap(merge.obj, features = top10$gene, group.by = "seurat_clusters") +
  theme(text = element_text(size = 10)) +
  scale_fill_gradientn(colors = c("blue", "white", "red")) 
ggsave("./Figures/Genes_Violin/Heatmap_top10.tiff", device='tiff', dpi=500, height = 5, width = 5, unit = 'in')
#This is a bit crowded, so let's select fewer genes (more signal, less noise)

#Top 3 genes per cluster by average log2FC
merge.obj.marker %>%
    group_by(cluster) %>%
    top_n(n = 3, wt = avg_log2FC) -> top3

top3 <- top3[c(1:9,16:18,10:15),]

DoHeatmap(merge.obj, features = top3$gene,group.by = "seurat_clusters")+
  theme(text = element_text(size = 14)) + 
  theme(axis.title.x = element_text(size = 8)) + #not axis text, not axis title, not plot title, not legend text, tried using a single theme() and two themes() - same result
  scale_fill_gradientn(colors = c("blue", "white", "red"))
ggsave("./Figures/Genes_Violin/Heatmap_top3.tiff", device='tiff', dpi=500, height = 5, width = 5, unit = 'in')

#Let's examine the genes with the highest average log2FC by cluster. Are they expressed primarily in that cluster?
VlnPlot(merge.obj, 
        features = c("Gm11549", "Dlk1", "Shox2", "Dsp", "Cabp7", "Adora2a"), 
        stack = T,
        flip = T)
#This produces good marker genes for clusters 2, 4, and 5.

#We are still lacking genes for clusters 1 & 3. Let's consider the other genes with high avg log2FC for clusters 1 and 3

VlnPlot(merge.obj, 
        features = c("Arhgap36", "Pmch",#1
                     "Ddn", "Ptgds"), #3 
        stack = T,
        flip = T)
#None of these are unique to their cluster.


#The following list of genes is from the FA project. In that analysis, these genes were effective marker genes by cluster, as they were found in one cluster but not others. That is not uniformly true for this analysis - some are present across clusters, others are not present in meaningful amounts in the merged adult and young dataset.

VlnPlot(merge.obj, 
        features = c("Ovol2", "Abhd12b","Thbs4","Ptpn14","Hcrt","Adora2a"), 
        stack = T,
        flip = T)
#This does not help with clusters 1 or 3. It is noteworthy that Adora2a was prominent in both the FA project and this analysis.

#Let's look at the top 10 genes by avg log2FC for cluster 1
VlnPlot(merge.obj, 
        features = c("Gpx3", "Ecel1", "Baiap3", "AW551984", "Dlk1", "Arhgap36", "Peg10", "Igsf1", "Cartpt", "Pmch"), 
        stack = T,
        flip = T)
ggsave("./Figures/Genes_Violin/vln_cluster1_top10_log2fc.tiff", device='tiff', dpi=500, height = 5, width = 5, unit = 'in')

#And now the top 10 genes by avg log2FC for cluster 3
VlnPlot(merge.obj, 
        features = c("Ddn","Camk2a","Cabp7","Gfap","Hpca","Wipf3","Nr3c2","Ptgds","Neurod6","Fn1"), #Cluster 3
        stack = T,
        flip = T)
ggsave("./Figures/Genes_Violin/vln_cluster1_top10_log2fc.tiff", device='tiff', dpi=500, height = 5, width = 5, unit = 'in')


```
### Step 2b - Marker Genes by Difference
As log2fc did not lead to uniform identification of marker genes, we next consider the difference in expression. FindAllMarkers reports pct.1 (the percentage this gene is expressed in this cluster) and pct.2 (the percentage it is expressed across all clusters). Let us examine if pct.1-pct.2 produces effective marker genes by cluster.

```{r}
merge.obj.marker$diff <- merge.obj.marker$pct.1 -merge.obj.marker$pct.2

merge.obj.marker %>%
    group_by(cluster) %>%
    top_n(n = 3, wt = diff) -> top3_diff

top3_diff <- top3_diff[c(1:9,16:18,10:15),]

DoHeatmap(merge.obj, features = top3_diff$gene,group.by = "seurat_clusters")+
  theme(text = element_text(size = 14)) + 
  theme(axis.title.x = element_text(size = 8)) + #not axis text, not axis title, not plot title, not legend text, tried using a single theme() and two themes() - same result
  scale_fill_gradientn(colors = c("blue", "white", "red"))
ggsave("./Figures/Genes_Violin/Heatmap_top3_diff.tiff", device='tiff', dpi=500, height = 5, width = 5, unit = 'in')


VlnPlot(merge.obj, 
        features = c("Gm11549", "Gpx3", "Tcf7l2", "Dsp", "Cabp7", "Adora2a"),
        stack = T,
        flip = T)
ggsave("./Figures/Genes_Violin/vln_top3_diff.tiff", device='tiff', dpi=500, height = 5, width = 5, unit = 'in')
#This did not produce good markers for clusters 1 or 3. 
#Let's examine the other top genes by difference for clusters 1 & 3.

#Cluster 1
VlnPlot(merge.obj, 
        features = c("Gpx3", "Ecel1", "Baiap3", "AW551984", "Dlk1", "Arhgap36", "Rassf8", "Peg10", "Agt", "Igsf1"), #These are the top genes from cluster 1 by difference
        stack = T,
        flip = T)
#None are unique - Clusters 1 & 2 seem related

#Cluster 3
VlnPlot(merge.obj, 
        features = c("Ddn","Cabp7","Gfap","Wipf3","Zbtb20","Nr3c2","Sowaha","Lhx2","Neurod6","Mertk"), #These are the top genes from cluster 3 by difference
        stack = T,
        flip = T)
#None are unique

#Clusters 1 and 3 correspond anatomically to the hippocampus, white matter regions, and hypothalamus. 
#Is there a biological explanation?

#Let's look at the hippocampal marker genes identified by Wu et al. (2024)
#HPF = hippocampal formation in their work
VlnPlot(merge.obj, 
        features = c("Pkp2", "Gpr161", "Homer3", "Dsc3", "Grp", "Prox1"), 
        stack = T,
        flip = T)


```
### Step 2c - Marker Genes by Anatomy

In the previous steps, we examined marker genes by cluster. For some clusters (especially cluster 3, to a lesser extent cluster 1), conventional approaches to finding marker genes were unsuccessful. Thus, we considered how the clusters relate to anatomical regions and combined clusters to cover the hippocampus. Simultaneously, anatomical names were assigned to each cluster for improved visualization and interpretation of results and considered a lower threshold for pct in order to maximize the difference between pct.1 and pct.2. 

```{r}
###Anatomical clustering
cluster.n <- seq(0,5,1)
anatomy <- c("Cortex", #0
              "Hypothalamus","Thalamus","Hippocampus","Striatum","Hippocampus")#1-5

cluster_an.df <- data.frame(cluster = cluster.n,
                            anatomy = anatomy)

merge.obj$anatomy <- cluster_an.df$anatomy[match(merge.obj$seurat_clusters,cluster_an.df$cluster)]
merge.obj$anatomy <- factor(merge.obj$anatomy, levels = c(c("Cortex","Hippocampus","Thalamus","Hypothalamus","Striatum")))

SpatialDimPlot(merge.obj, images = c("adult_1", "adult_2"), pt.size.factor = 3, group.by = "anatomy", ncol = 2)
SpatialDimPlot(merge.obj, images = c("mf_p21_rep1", "mf_p21_rep2"), pt.size.factor = 740, group.by = "anatomy", ncol = 2)

SpatialDimPlot(merge.obj, images = "adult_1",pt.size.factor = 3,group.by = "seurat_clusters",ncol = 1)
#To compare the previous clustering result to current clustering by anatomy

###Marker genes by anatomy
Idents(merge.obj) <- "anatomy"
#Top 10 genes per cluster by anatomy
merge.obj.marker.an <- FindAllMarkers(merge.obj, only.pos = TRUE, min.pct = 0.1, logfc.threshold = 0.25)

merge.obj.marker.an$pct.diff <- merge.obj.marker$pct.1 - merge.obj.marker.an$pct.2

#Repeat Heatmap
merge.obj.marker.an %>%
    group_by(cluster) %>%
    top_n(n = 3, wt = avg_log2FC) -> top3_an

DoHeatmap(merge.obj, features = top3_an$gene, group.by = "anatomy") +
  theme(text = element_text(size = 10)) +
  scale_fill_gradientn(colors = c("blue", "white", "red"))# + theme(axis.text.y = element_blank())
# ggsave("./Figures/Genes_Violin/Heatmap_top3_an.tiff", device='tiff', dpi=500, height = 5, width = 5, unit = 'in')

```
### Step 2d - Marker Genes by Anatomical Region
Manually, we examined the top 10 dataset for genes with a high pct.diff as well as low pct.2. The best results are shared below. 

Liduo shared that it is acceptable for the marker gene to be present in two clusters, for example cluster N and cluster M as long as either cluster N or M has its unique marker gene. If cluster N has a unique marker gene, then a shared marker gene for N and M can be used to isolate cluster M. 

This is done with the cortex & hippocampus, as well as thalamus & hypothalamus below. 

```{r}
VlnPlot(merge.obj, 
        features = c("Emx1","Crlf1", 
                     "Arhgap36",
                     "Grid2ip","Adora2a"), #3 
        stack = T,
        flip = T)

```
### Step 3 - Volcano plots
Next, let's consider differential gene expression, visualized by volcano plots.


```{r}
#Create a new metadata column to store age data.
merge.obj$age <- 0
merge.obj$age[merge.obj$orig.ident %in% c("mf_p21_rep1","mf_p21_rep2")] <- "P21"
merge.obj$age[merge.obj$orig.ident %in% c("adult_1","adult_2")] <- "Adult"

Idents(merge.obj) <- "anatomy"

#Loop to make old to young mice in each brain region
for(br in unique(merge.obj$anatomy)){
      temp.obj <- subset(merge.obj, anatomy==br)
marker <- FindMarkers(temp.obj,
                      ident.1 = "P21", ident.2 = "Adult",
                      group.by = 'age'
                      )
      
      marker$brain.region <- br
      
      marker$gene <- rownames(marker)
      marker$regulation <- ifelse(marker$avg_log2FC > 0,paste0("Up"),paste0("Down"))

      marker$sig <- "not_sig"
      marker$sig[abs(marker$avg_log2FC) > log2(1.2) & marker$p_val_adj < 0.05] <- paste(br,"sig",sep = "_")
      marker$sig <- factor(marker$sig, levels = c("not_sig",paste(br,"sig",sep = "_")))
      
      assign(paste(br,"deg",sep = "."), marker)
      }

#Preparation for volcano plot
combine.deg.df <- rbind(Cortex.deg,
                        Hippocampus.deg,
                        Hypothalamus.deg,
                        Thalamus.deg,
                        Striatum.deg)
colnames(combine.deg.df)[6] <- "cluster"

#VolcanoPlot
scRNAtoolVis::markerVolcano(markers = combine.deg.df,
              topn = 10,
              labelCol = ggsci::pal_npg()(9))
```
### Step 4a - Gene Ontology (GO)

We have identified which genes are upregulated or downregulated in each cluster. Next, we would like to identify what is the general purpose of each gene and what purposes occur most frequently by cluster. 

First, we will examine the genes upregulated in p21 mice.

```{r}
#this will analyze the genes upregulated in p21 mice.
assign("eg", bitr(rownames(marker), fromType="SYMBOL", toType="ENTREZID",OrgDb="org.Mm.eg.db"))

      ego <- enrichGO(gene = eg$ENTREZID,
                      OrgDb = org.Mm.eg.db,
                      ont = "BP",
                      pAdjustMethod = "BH",
                      pvalueCutoff  = 0.05,
                      qvalueCutoff = 0.05,
                      readable = TRUE)
      if (nrow(ego@result[ego@result$p.adjust < 0.05 & ego@result$qvalue < 0.05,]) <= 0){
        next
      }
      ego <- pairwise_termsim(ego)
      ego <- simplify(ego)
      
      time<-c("P21","Adult")
      
      # write.csv(ego@result, file = paste(ct.name,time[i],"cfcvscc.GO.csv", sep = "."))
      write.csv(ego@result, file = paste(br,time[br],"hipweanedvsadult_20241003.GO.csv", sep = "."))
      # write.csv(ego@result, file = paste(ct.name,time[i],"ccvshomecage.GO.csv", sep = "."))

      go_analysis <- ego@result[ego@result$p.adjust < 0.05 & ego@result$qvalue < 0.05,]

      simMatrix <- calculateSimMatrix(go_analysis$ID,
                                      orgdb="org.Mm.eg.db",
                                      ont="BP",
                                      method="Rel")

      scores <- setNames(-log10(go_analysis$qvalue), go_analysis$ID)
      reducedTerms <- reduceSimMatrix(simMatrix,
                                      scores,
                                      threshold=0.7,
                                      orgdb="org.Mm.eg.db")
      treemapPlot(reducedTerms, position.legend = "right", fontsize.legend = 6) 

# #      ggsave(paste(br,time[br],"weanedvsadult.GO.tiff", sep = "."),
#              device='tiff', dpi=500, height = 10, width = 10, unit = 'in')
#       ggsave(paste(br,time[br],"weanedvsadult.GO.csv", sep = "."),
#              device='tiff', dpi=500, height = 10, width = 10, unit = 'in')

#Let's also add a dotplot
dotplot(ego, showCategory=20)
#ggsave('Figures/dotplot.tiff', device='tiff', dpi=500, 
#       height = 8, width = 8, unit = 'in')

heatplot_image <- heatplot(ego, showCategory=7)
# ggsave('Figures/heatplot2.tiff', device='tiff', dpi=500,
#        height = 8, width = 8, unit = 'in')
#That is not useful. The genes all overlap and are not legible.

barplot(ego, showCategory=7)
#ggsave('Figures/barplot.tiff', device='tiff', dpi=500,
#       height = 8, width = 8, unit = 'in')

#Supplemental figure 5 in FA
#upsetplot(ego)
#That is not useful. The genes all overlap and are not legible.

```

###Step 5a - Hippocampus - Upregulated in p21
First, let's view the gene annotation sets upregulated in the p21 mouse hippocampus. 

```{r}
#Let's try to only view the genes upregulated in the hippocampus in p21 brains
hip_data <- subset(merge.obj, anatomy=="Hippocampus")
hip_marker <- FindMarkers(hip_data,
                      ident.1 = "P21", ident.2 = "Adult",
                      group.by = 'age'
                      )
      
      hip_marker$brain.region <- "Hippocampus"
      
      hip_marker$gene <- rownames(hip_marker)
      hip_marker$regulation <- ifelse(hip_marker$avg_log2FC > 0,paste0("Up"),paste0("Down"))
      hip_marker <- subset(hip_marker, hip_marker$avg_log2FC >0)
      #only the upregulated genes
        
        
      hip_marker$sig <- "not_sig"
      hip_marker$sig[abs(hip_marker$avg_log2FC) > log2(1.2) & hip_marker$p_val_adj < 0.05] <- paste("Hippocampus","sig",sep = "_")
      hip_marker$sig <- factor(hip_marker$sig, levels = c("not_sig",paste("Hippocampus","sig",sep = "_")))

length(unique(hip_marker$gene))      
#10136 genes upregulated in p21 mice in the hippocampus
          
assign("eg", bitr(rownames(hip_marker), fromType="SYMBOL", toType="ENTREZID",OrgDb="org.Mm.eg.db"))

      hip_ego <- enrichGO(gene = eg$ENTREZID,
                      OrgDb = org.Mm.eg.db,
                      ont = "BP",
                      pAdjustMethod = "BH",
                      pvalueCutoff  = 0.05,
                      qvalueCutoff = 0.05,
                      readable = TRUE)
      if (nrow(hip_ego@result[hip_ego@result$p.adjust < 0.05 & hip_ego@result$qvalue < 0.05,]) <= 0){
        next
      }
      hip_ego <- pairwise_termsim(hip_ego)
      hip_ego <- simplify(hip_ego)

length(unique(hip_ego@result[["Description"]]))      
#699 unique GO annotation terms

      time<-c("P21","Adult")
      
      # write.csv(ego@result, file = paste(ct.name,time[i],"cfcvscc.GO.csv", sep = "."))
      write.csv(hip_ego@result, file = paste("Hippocampus","hipweanedvsadult_20241003.GO.csv", sep = "."))        

#This produces 891 unique GO annotation terms. Let's group similar terms for more interpretable results
      hip_go_analysis <- hip_ego@result[hip_ego@result$p.adjust < 0.05 & ego@result$qvalue < 0.05,]
      hip_simMatrix <- calculateSimMatrix(hip_go_analysis$ID,
                                      orgdb="org.Mm.eg.db",
                                      ont="BP",
                                      method="Rel")

      hip_scores <- setNames(-log10(hip_go_analysis$qvalue), hip_go_analysis$ID)
      hip_reducedTerms9 <- reduceSimMatrix(hip_simMatrix,
                                      hip_scores,
                                      threshold=0.9,#increase threshold to reduce # of groups
                                      orgdb="org.Mm.eg.db")
#0.9 gives 34 clusters instead of 0.7 giving 137; 0.95 gives 32
#but there are multiple terms associated per cluster
write.csv(hip_reducedTerms9, file = paste("hip_UP_reducedTerms9.csv", sep = ","))
#collect a list of only the unique parent terms
hip_up_reducedTerms9_parentTerms <- c()
hip_up_reducedTerms9_parentTerms <- unique(hip_reducedTerms9$parentTerm)

print(hip_up_reducedTerms9_parentTerms)

#  [1] "extracellular matrix organization"                                     
#  [2] "ameboidal-type cell migration"                                             
#  [3] "epithelial tube morphogenesis"                                             
#  [4] "cell-substrate adhesion"                                                   
#  [5] "phospholipid metabolic process"                                            
#  [6] "tissue migration"                                                          
#  [7] "negative regulation of locomotion"                                         
#  [8] "regulation of epithelial cell proliferation"                               
#  [9] "regulation of cell growth"                                                 
# [10] "calcium ion transport"                                                     
# [11] "regulation of cellular response to growth factor stimulus"                 
# [12] "positive regulation of hydrolase activity"                                 
# [13] "cell-cell signaling by wnt"                                                
# [14] "regulation of actin cytoskeleton organization"                             
# [15] "neuron apoptotic process"                                                  
# [16] "regulation of cellular component size"                                     
# [17] "rhythmic process"                                                          
# [18] "establishment or maintenance of cell polarity"                             
# [19] "regulation of microtubule-based process"                                   
# [20] "sex differentiation"                                                       
# [21] "temperature homeostasis"                                                   
# [22] "chromosome segregation"                                                    
# [23] "regulation of cell division"                                               
# [24] "pigmentation"                                                              
# [25] "B cell receptor signaling pathway"                                         
# [26] "biological process involved in interaction with host"                      
# [27] "cell recognition"                                                          
# [28] "single stranded viral RNA replication via double stranded DNA intermediate"
# [29] "membrane docking"                                                          
# [30] "regulation of plasma lipoprotein particle levels"                          
# [31] "cell-cell fusion"                                                          
# [32] "biological phase" 


#Let's visualize the parent terms
# heatmapPlot(  #package rrvgo
#   hip_simMatrix,
#   reducedTerms = hip_reducedTerms9,
#   annotateParent = TRUE,
#   annotationLabel = "parentTerm",
#   fontsize=6
# )
#Everything is overlapped;unhelpful


dotplot(hip_ego, showCategory=18)
ggsave('Figures/Hip/up_hip_dotplot.tiff', device='tiff', dpi=500,
      height = 8, width = 8, unit = 'in')

#heatplot_image <- heatplot(ego, showCategory=10)
# ggsave('Figures/heatplot2.tiff', device='tiff', dpi=500,
#        height = 8, width = 8, unit = 'in')
#That is not useful. The genes all overlap and are not legible.

barplot(hip_ego, showCategory=18)
ggsave('Figures/Hip/up_hip_barplot.tiff', device='tiff', dpi=500,
      height = 8, width = 8, unit = 'in')
#Let's arrange it to display Count in descending order
sorted_hip_ego <- hip_ego %>%
  arrange(desc(hip_ego@result[["Count"]]))
barplot(sorted_hip_ego, showCategory=18, font = 12)
ggsave('Figures/Hip/up_hip_barplot.tiff', device='tiff', dpi=500,
      height = 8, width = 8, unit = 'in')
      
# ####
# #make a matrix for the heatmap
# hip_ego_description <- hip_ego@result[["Description"]]
# hip_ego_zscore <- hip_ego@result[["zScore"]]
# #hip_ego_subset_df <- data.frame(Description = hip_ego_description, zScore = hip_ego_zscore)
# hip_ego_subset_mx <- as.matrix(hip_ego_zscore)
# rownames(hip_ego_subset_mx) <- hip_ego_description
# 
# pheatmap(hip_ego_subset_mx, scale = "row")
# #heat map should be for all brain regions - columns are regions, rows are GO annotation terms      
```

###Step 5b - Hippocampus - Upregulated in Adults
Next, let's examine which genes are upregulated in adults (i.e. downregulated in p21)

```{r}
hip_data <- subset(merge.obj, anatomy=="Hippocampus")
hip_marker <- FindMarkers(hip_data,
                      ident.1 = "P21", ident.2 = "Adult",
                      group.by = 'age'
                      )
      
      hip_marker$brain.region <- "Hippocampus"
      
      hip_marker$gene <- rownames(hip_marker)
      hip_marker$regulation <- ifelse(hip_marker$avg_log2FC > 0,paste0("Up"),paste0("Down"))
      Dhip_marker <- subset(hip_marker, hip_marker$avg_log2FC < 0)
      #only the downregulated genes
        
        
      Dhip_marker$sig <- "not_sig"
      Dhip_marker$sig[abs(Dhip_marker$avg_log2FC) > log2(1.2) & Dhip_marker$p_val_adj < 0.05] <- paste("Hippocampus","sig",sep = "_")
      Dhip_marker$sig <- factor(Dhip_marker$sig, levels = c("not_sig",paste("Hippocampus","sig",sep = "_")))

length(unique(Dhip_marker$gene))      
# 4939 unique marker genes that were downregulated in p21 mice in the hippocampus      
      
      
          
assign("eg", bitr(rownames(Dhip_marker), fromType="SYMBOL", toType="ENTREZID",OrgDb="org.Mm.eg.db"))

      Dhip_ego <- enrichGO(gene = eg$ENTREZID,
                      OrgDb = org.Mm.eg.db,
                      ont = "BP",
                      pAdjustMethod = "BH",
                      pvalueCutoff  = 0.05,
                      qvalueCutoff = 0.05,
                      readable = TRUE)
      if (nrow(Dhip_ego@result[Dhip_ego@result$p.adjust < 0.05 & Dhip_ego@result$qvalue < 0.05,]) <= 0){
        next
      }
      Dhip_ego <- pairwise_termsim(Dhip_ego)
      Dhip_ego <- simplify(Dhip_ego)

length(unique(Dhip_ego@result[["Description"]]))
#688 unique GO terms upregulated in adults 
      time<-c("P21","Adult")
      
      # write.csv(ego@result, file = paste(ct.name,time[i],"cfcvscc.GO.csv", sep = "."))
      write.csv(Dhip_ego@result, file = paste("Hippocampus","DOWNhipweanedvsadult.GO.csv", sep = "."))        

#This produces _____ unique GO annotation terms. Let's group similar terms for more interpretable results
      Dhip_go_analysis <- Dhip_ego@result[Dhip_ego@result$p.adjust < 0.05 & Dhip_ego@result$qvalue < 0.05,]
      Dhip_simMatrix <- calculateSimMatrix(Dhip_go_analysis$ID,
                                      orgdb="org.Mm.eg.db",
                                      ont="BP",
                                      method="Rel")

      Dhip_scores <- setNames(-log10(Dhip_go_analysis$qvalue), Dhip_go_analysis$ID)
      Dhip_reducedTerms9 <- reduceSimMatrix(Dhip_simMatrix,
                                      Dhip_scores,
                                      threshold=0.9,#increase threshold to reduce # of groups
                                      orgdb="org.Mm.eg.db")

write.csv(Dhip_reducedTerms9, file = paste("hip_DOWN_reducedTerms9.csv", sep = ","))
#collect a list of only the unique parent terms
hip_down_reducedTerms9_parentTerms <- c()
hip_down_reducedTerms9_parentTerms <- unique(Dhip_reducedTerms9$parentTerm)

print(hip_down_reducedTerms9_parentTerms)

#  [1] "small GTPase-mediated signal transduction"                      
#  [2] "establishment of protein localization to organelle"             
#  [3] "postsynapse organization"                                       
#  [4] "negative regulation of phosphorus metabolic process"            
#  [5] "regulation of membrane potential"                               
#  [6] "learning or memory"                                             
#  [7] "regulation of monoatomic ion transmembrane transporter activity"
#  [8] "mitotic cell cycle phase transition"                            
#  [9] "immune response-regulating signaling pathway"                   
# [10] "intracellular calcium ion homeostasis"                          
# [11] "rhythmic process"                                               
# [12] "ameboidal-type cell migration"                                  
# [13] "regulation of cell growth"                                      
# [14] "neuron apoptotic process"                                       
# [15] "forebrain development"                                          
# [16] "muscle cell proliferation"                                      
# [17] "microtubule-based movement"                                     
# [18] "Wnt signaling pathway"                                          
# [19] "regulation of actin filament-based process"                     
# [20] "regulation of leukocyte cell-cell adhesion"                     
# [21] "viral process"                                                  
# [22] "negative regulation of locomotion"                              
# [23] "modulation by host of symbiont process"                         
# [24] "establishment or maintenance of cell polarity"                  
# [25] "social behavior"                                                
# [26] "stem cell division"                                             
# [27] "organelle localization by membrane tethering"                   
# [28] "regulation of reproductive process"                             
# [29] "regulation of plasma lipoprotein particle levels"               
# [30] "developmental pigmentation"   

dotplot(Dhip_ego, showCategory=18)
ggsave('Figures/Hip/down_hip_dotplot.tiff', device='tiff', dpi=500,
      height = 8, width = 8, unit = 'in')

#heatplot_image <- heatplot(ego, showCategory=10)
# ggsave('Figures/heatplot2.tiff', device='tiff', dpi=500,
#        height = 8, width = 8, unit = 'in')
#That is not useful. The genes all overlap and are not legible.


#Let's arrange the bar plot to display Count in descending order
sorted_Dhip_ego <- Dhip_ego %>%
  arrange(desc(Dhip_ego@result[["Count"]]))
barplot(sorted_Dhip_ego, showCategory=18, font = 12)
ggsave('Figures/Hip/down_hip_barplot.tiff', device='tiff', dpi=500,
      height = 8, width = 8, unit = 'in')

#Let's explore some other visualization options, such as a treeplot.
treemapPlot(Dhip_reducedTerms9, position.legend = "right", fontsize.legend = 6)
ggsave('Figures/Hip/down_hip_treemapplot.tiff', device='tiff', dpi=500,
      height = 8, width = 8, unit = 'in')
#Figure is too small

#How many parent terms are shared by p21 and adult?
parent_terms_Dhip <- unique(Dhip_reducedTerms9$parentTerm)
parent_terms_hip <- unique(hip_reducedTerms9$parentTerm)
hip_common_terms <- intersect(parent_terms_Dhip, parent_terms_hip)
print(hip_common_terms)

#goplot(Dhip_ego)
#Error in AnnotationDbi::mget(id, GOANCESTOR) : 
#  second argument must be an environment

#wordcloudPlot(hip_reducedTerms9)
```

